### ISSUE TITLE
CPU memory leak when running train_yolov3.py

### ISSUE THREAD

**Issue by houweidong** [apache/incubator-mxnet#14742]
When i run the train_yolo3.py in gluoncv, the cpu memory occupation is about 21G at epoch 0,  but at the end of every epoch, the cpu memory will increase about 1G~2G. And the cpu memory can hit 50G when the epochs is about 40!
How can i solve this porblem?

train with option: 
`--batch-size 16    -j 8   --dataset coco  --syncbn`

version:
```
ubuntu 18.04
mxnet   (1.5.0b20190411  and 1.5.0b20180305)
gluoncv (0.5.0b20190411  and 0.4.0b20190227)
10 1080ti
cuda 9.0
```

**Comment by frankfliu** [comment]
@mxnet-label-bot add [Memory]

**Comment by lanking520** [comment]
@zhreshold

**Comment by zhreshold** [comment]
@houweidong 
can you paste the memory consumption by pid using `ps aux --sort -rss | head -n` ?
Without detailed info I have no idea what's wrong with it.

**Comment by houweidong** [comment]
ps aux --sort -rss | head -20
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      7551  436 13.3 134231992 13047996 pts/0 Sl+ 11:28 122:55 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8016 93.1  7.5 12621720 7394328 pts/0 Sl+ 11:29  25:22 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7967 91.7  7.5 12631112 7379188 pts/0 Sl+ 11:29  25:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8034 91.9  7.5 12604824 7378900 pts/0 Sl+ 11:29  25:02 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7988 92.0  7.5 12605612 7378328 pts/0 Sl+ 11:29  25:05 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7997 91.9  7.5 12604048 7377076 pts/0 Rl+ 11:29  25:03 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8025 92.1  7.5 12573304 7345996 pts/0 Sl+ 11:29  25:06 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7979 92.1  7.5 12570536 7344564 pts/0 Sl+ 11:29  25:06 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8007 92.4  7.5 12562104 7334816 pts/0 Sl+ 11:29  25:11 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8076  0.0  7.4 12646640 7315760 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8085  0.0  7.4 12646640 7315760 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8094  0.0  7.4 12646640 7315760 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8103  0.0  7.4 12646640 7315760 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8115  0.0  7.4 12646640 7315760 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8058  0.0  7.4 12646640 7315756 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8067  0.0  7.4 12646640 7315756 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8049  0.0  7.4 12646640 7315752 pts/0 Sl+ 11:29   0:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root     10965  2.5  1.5 12241692 1481400 pts/0 Sl+ 4月05 639:09 /root/pycharm/jre64/bin/java -classpath /root/pycharm/lib/bootstrap.jar:/root/pycharm/lib/extensions.jar:/root/pycharm/lib/util.jar:/root/pycharm/lib/jdom.jar:/root/pycharm/lib/log4j.jar:/root/pycharm/lib/trove4j.jar:/root/pycharm/lib/jna.jar -Xms128m -Xmx750m -XX:ReservedCodeCacheSize=240m -XX:+UseConcMarkSweepGC -XX:SoftRefLRUPolicyMSPerMB=50 -ea -Dsun.io.useCanonCaches=false -Djava.net.preferIPv4Stack=true -Djdk.http.auth.tunneling.disabledSchemes="" -XX:+HeapDumpOnOutOfMemoryError -XX:-OmitStackTraceInFastThrow -Dawt.useSystemAAFontSettings=lcd -Dsun.java2d.renderer=sun.java2d.marlin.MarlinRenderingEngine -XX:ErrorFile=/root/java_error_in_PYCHARM_%p.log -XX:HeapDumpPath=/root/java_error_in_PYCHARM.hprof -Didea.paths.selector=PyCharmCE2018.3 -Djb.vmOptionsFile=/root/pycharm/bin/pycharm64.vmoptions -Didea.platform.prefix=PyCharmCore com.intellij.idea.Main
erised   35435  5.0  0.9 3785860 974604 ?      Sl   4月09 993:36 /opt/teamviewer/tv_bin/TeamViewer_Desktop

```
this is at epoch 5, CPUmemory occupation is about 30G

......................................................................................................................................................................................................................
......................................................................................................................................................................................................................
......................................................................................................................................................................................................................


ps aux --sort -rss | head -20
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      7551  438 13.5 129978316 13223692 pts/0 Sl+ 11:28 244:38 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7997 90.6  7.5 12600784 7374952 pts/0 Sl+ 11:29  49:44 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8025 91.1  7.5 12590040 7364248 pts/0 Sl+ 11:29  50:00 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7967 90.7  7.5 12638704 7356276 pts/0 Sl+ 11:29  49:47 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7979 91.3  7.5 12573112 7346988 pts/0 Sl+ 11:29  50:06 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8034 90.7  7.5 12572472 7346384 pts/0 Sl+ 11:29  49:47 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      7988 90.8  7.5 12618464 7346056 pts/0 Sl+ 11:29  49:52 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8007 90.9  7.5 12571272 7345172 pts/0 Sl+ 11:29  49:54 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8016 91.3  7.5 12570272 7344144 pts/0 Sl+ 11:29  50:09 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8094  0.3  7.5 12646896 7325600 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8103  0.3  7.5 12646896 7325600 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8049  0.3  7.5 12646896 7325596 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8058  0.3  7.5 12646896 7325596 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8085  0.3  7.5 12646896 7325596 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8115  0.3  7.5 12646896 7325596 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8076  0.2  7.5 12646896 7325592 pts/0 Sl+ 11:29   0:09 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root      8067  0.3  7.5 12646896 7325588 pts/0 Sl+ 11:29   0:10 /usr/bin/python3.5 /root/models/yolov3_origin/train_yolo3.py --gpus 1,2 --network darknet53 --syncbn --batch-size 16 -j 8 --val-interval 10
root     10965  2.5  1.5 12241692 1482248 pts/0 Sl+ 4月05 640:07 /root/pycharm/jre64/bin/java -classpath /root/pycharm/lib/bootstrap.jar:/root/pycharm/lib/extensions.jar:/root/pycharm/lib/util.jar:/root/pycharm/lib/jdom.jar:/root/pycharm/lib/log4j.jar:/root/pycharm/lib/trove4j.jar:/root/pycharm/lib/jna.jar -Xms128m -Xmx750m -XX:ReservedCodeCacheSize=240m -XX:+UseConcMarkSweepGC -XX:SoftRefLRUPolicyMSPerMB=50 -ea -Dsun.io.useCanonCaches=false -Djava.net.preferIPv4Stack=true -Djdk.http.auth.tunneling.disabledSchemes="" -XX:+HeapDumpOnOutOfMemoryError -XX:-OmitStackTraceInFastThrow -Dawt.useSystemAAFontSettings=lcd -Dsun.java2d.renderer=sun.java2d.marlin.MarlinRenderingEngine -XX:ErrorFile=/root/java_error_in_PYCHARM_%p.log -XX:HeapDumpPath=/root/java_error_in_PYCHARM.hprof -Didea.paths.selector=PyCharmCE2018.3 -Djb.vmOptionsFile=/root/pycharm/bin/pycharm64.vmoptions -Didea.platform.prefix=PyCharmCore com.intellij.idea.Main
erised   35435  5.0  1.0 3785860 990668 ?      Sl   4月09 999:05 /opt/teamviewer/tv_bin/TeamViewer_Desktop

```
this is at epoch 10, CPUmemory occupation is about 40G

**Comment by houweidong** [comment]
i add code in train_yolov3.py at every epoch as below to get result quickly.
```
test = 0
for i, batch in enumerate(train_data):
    test += 1
    if test > 500:
        break
```

**Comment by zhreshold** [comment]
Looks like the men occupied by mxnet main thread and all workers are constant.
I highly suspect that it's the shared memory issue.
What do you get from `df -h /dev/shm` ?

**Comment by houweidong** [comment]
in the docker contrainer:
df -h /dev/shm:
```
Filesystem      Size  Used Avail Use% Mounted on
tmpfs            47G     0   47G   0% /dev/shm
```

Need i run this command at running time?

**Comment by zhreshold** [comment]
I cannot reproduce the problem, training yolo3, shared memory was consistent at 28G no matter how many epochs.

Can you try different versions of mxnet? 1.4.0 and some older nightly build?

**Comment by houweidong** [comment]
mxnet 1.4.0 has the same problem
the train_yolov3.py dosen't support the mxnet 1.3.0

and the problem seems not the hardware leads to. In single 1080ti, 10 1080ti or 9 2080ti server, the CPU memory leak problem exist.

**Comment by zhreshold** [comment]
@houweidong need further investigation, please use thread_pool=True to avoid memory issue as a temporary solution.

**Comment by houweidong** [comment]
i just add thread_pool in three place: ("n" presents the nth added param )

```
def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_workers, args):
    """Get dataloader."""
    width, height = data_shape, data_shape
    batchify_fn = Tuple(*([Stack() for _ in range(6)] + [Pad(axis=0, pad_val=-1) for _ in range(1)]))  # stack image, all targets generated
    if args.no_random_shape:
        train_loader = gluon.data.DataLoader(
            train_dataset.transform(YOLO3DefaultTrainTransform(width, height, net, mixup=args.mixup)),
            batch_size, True, batchify_fn=batchify_fn, last_batch='rollover', num_workers=num_workers, thread_pool=True  "1")
    else:
        transform_fns = [YOLO3DefaultTrainTransform(x * 32, x * 32, net, mixup=args.mixup) for x in range(10, 20)]
        train_loader = RandomTransformDataLoader(
            transform_fns, train_dataset, batch_size=batch_size, interval=10, last_batch='rollover',
            shuffle=True, batchify_fn=batchify_fn, num_workers=num_workers, thread_pool=True  "2")
    val_batchify_fn = Tuple(Stack(), Pad(pad_val=-1))
    val_loader = gluon.data.DataLoader(
        val_dataset.transform(YOLO3DefaultValTransform(width, height)),
        batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers, thread_pool=True  "3")
    return train_loader, val_loader
```

and the problem exits still(25 epoch error, should be the full shm leads to)

**Comment by zrss** [comment]
anything new ? i encouter `Inconsistency detected by ld.so: dl-tls.c: 493: _dl_allocate_tls_init: Assertion `listp->slotinfo[cnt].gen <= GL(dl_tls_generation)' failed!` ...

**Comment by chinakook** [comment]
I've also met with this problem when runing train_faster_rcnn.py(gluon-cv) without syncbn. Set ```thread_pool=True``` can avoid this problem.

**Comment by chinakook** [comment]
I found the reason about train_faster_rcnn.py memory leak. https://github.com/apache/incubator-mxnet/issues/19574
